src/application_default_credentials.rs

```diff
@@ -0,0 +1,48 @@
+use crate::error::Error;
+use crate::types::TokenInfo;
+
+/// Provide options for the Application Default Credential Flow, mostly used for testing
+pub struct ApplicationDefaultCredentialsFlowOpts {
+    /// Used as base to build the url during token request from GCP metadata server
+    pub metadata_url: Option<String>,
+}
+impl Default for ApplicationDefaultCredentialsFlowOpts {
+    fn default() -> Self {
+        Self { metadata_url: None }
+    }
+}
+
+pub struct ApplicationDefaultCredentialsFlow {
+    metadata_url: String,
+}
+
+impl ApplicationDefaultCredentialsFlow {
+    pub(crate) fn new(opts: ApplicationDefaultCredentialsFlowOpts) -> Self {
+        let metadata_url = opts.metadata_url.unwrap_or_else(|| "http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/token".to_string());
+        ApplicationDefaultCredentialsFlow { metadata_url }
+    }
+
+    pub(crate) async fn token<C, T>(
+        &self,
+        hyper_client: &hyper::Client<C>,
+        scopes: &[T],
+    ) -> Result<TokenInfo, Error>
+    where
+        T: AsRef<str>,
+        C: hyper::client::connect::Connect + Clone + Send + Sync + 'static,
+    {
+        let scope = crate::helper::join(scopes, ",");
+        let token_uri = format!("{}?scopes={}", self.metadata_url, scope); // TODO: This feels jank, can it be done better?
```

Comment:
One approach would be to parse the metadata URL early in into a rich URL type (e.g. https://docs.rs/url/2.2.2/url/), store that rich type in `metadata_url`, and then use URL methods here to add the scope. An advantage would be that malformed URLs would be diagnosed earlier.

Reply:
would require an additional dependency, and I do not expect anyone to actually supply custom metadata servers besides our tests.

Reply:
Fair enough.